### ISSUE TITLE
Add support for bound instruments to the metrics API

### ISSUE THREAD

**Issue by jack-berg** [open-telemetry/opentelemetry-specification#4126]
Bound instruments is the idea of obtaining a reference "bound" to a particular set of attributes which is known ahead of time. By doing so, you can record measurements directly to the bound reference and avoid an internal lookup.

For example, suppose I know ahead of time that I need to record measurements to attribute sets `{"color": "red", "shape", "square"}`, `{"color": "red", "shape", "circle"}`, `{"color": "blue", "shape", "square"}`, `{"color": "blue", "shape", "circle"}`.

Without bound instruments, code to record these series might look like:
```
// Initialize counter and known attribute sets
LongCounter counter = meter.counterBuilder("my-counter").ofLongs().build();

Attributes redSquare = Attributes.builder().put("color", "red").put("shape", "square").build();
Attributes redCircle = Attributes.builder().put("color", "red").put("shape", "circle").build();
Attributes blueSquare = Attributes.builder().put("color", "blue").put("shape", "square").build();
Attributes blueCircle = Attributes.builder().put("color", "blue").put("shape", "circle").build();

// ... record measurements
counter.increment(1, redSquare);
counter.increment(1, redCircle);
counter.increment(1, blueSquare);
counter.increment(1, blueCircle);
```
Each time a measurement is recorded, the internals of the SDK must lookup the aggregated state corresponding to the attributes in a map.

With bound instruments, you can avoid this map lookup and improve the performance of record operations:
```
// Initialize counter and obtain bound references to known attribute sets
LongCounter counter = meter.counterBuilder("my-counter").ofLongs().build();

BoundLongCounter redSquare = counter.bind(Attributes.builder().put("color", "red").put("shape", "square").build());
BoundLongCounter redCircle = counter.bind(Attributes.builder().put("color", "red").put("shape", "circle").build());
BoundLongCounter blueSquare = counter.bind(Attributes.builder().put("color", "blue").put("shape", "square").build());
BoundLongCounter blueCircle = counter.bind(Attributes.builder().put("color", "blue").put("shape", "circle").build());

// ... record measurements
redSquare.increment(1);
redCircle.increment(1);
blueSquare.increment(1);
blueCircle.increment(1);
```

We should consider extending the metrics API to support this concept. I recently wrote a [blog post](https://opentelemetry.io/blog/2024/java-metric-systems-compared/) comparing otel java metrics to other metrics systems in the java ecosystem (prometheus and micrometer), and one notable negative of otel was the lack of support for bound instruments. For java, avoiding the map lookup saves ~11 ns per record call on my machine.

Its worth noting that bound instruments serve a rather niche use case where: 1. The attribute sets are known ahead of time. 2. The application is very performance sensitive. If either of these is not true, then bound instruments are of limited value. (I previously argued that we should punt on support for bound instruments in the initial metrics API release for this very reason.) For example, some of the most popular metrics in otel are `http.server.request.duration` and `http.client.request.duration`, and neither of these would be able to benefit from bound instruments because the attribute values are computed at runtime using application context.

**Comment by yurishkuro** [comment]
I like the idea of bound instruments, but I would first be interested to understand how much of an improvement they would provide. You mentioned map lookup, but even in your concrete example you would either have some switch statement to know which counter to bump, or in more realistic situation you might have a map lookup in the user code anyway. For example, when Jaeger collector emits metrics about the volume of received spans, it tags them with `service.name`, so even though we use "bound" instrument via PromSDK, we still have the map lookup in our own code.

Having said that, there are certainly situations where labels are contextually known, e.g. RPC metrics by the endpoint - each endpoint implementation can have a reference to a bound instrument with no additional lookup.

What actually worries me is the implementation details below the surface. I was looking at Go SDK today and it goes to lengths to avoid memory allocations by allowing users to pass AttributeSet which has a precomputed hash code for efficient lookup. But then something else causes memory allocation which destroys the benchmark (https://github.com/jaegertracing/jaeger/pull/5676).

So what I would really like to understand is whether bound instruments can provide a path to allocation-free (and ideally lock free) counter bumps, and whether the same can be achieved with the existing API.

**Comment by cijothomas** [comment]
> avoiding the map lookup saves ~11 ns per record call on my machine

OTel .NET's benchmarks show a totally diff. story. With 0 attributes, it takes 10-15ns to perform a Counter.Add. With 1 attribute, it moves to 50, and with 5 attributes its near 200ns and so on. Bound instrument would give the same performance irrespective of the attribute count, and that would be the same as that of the 0 attributes case (.NET special cases the 0 attribute case, giving us a hint into the possible gains with bound instrument!)

In short, with bound instruments, I expect the bound_counter.Add call to take 10-15ns, irrespective of the attributes bound to it. Without bound, this could be several 100s, so the perf gains are much more.

(Used numbers from OTel .NET, and I see similar from OTel Rust too).

Also adding the slack conversation earlier this year: https://cloud-native.slack.com/archives/C01NP3BV26R/p1700165373479809?thread_ts=1700152641.765839&cid=C01NP3BV26R

**Comment by cijothomas** [comment]
@yurishkuro 

>  but I would first be interested to understand how much of an improvement they would provide

https://github.com/open-telemetry/opentelemetry-specification/issues/4126#issuecomment-2207011779 Shows expected improvement based on my experiments with OTel .NET, Rust. I don't think it'll be significantly diff. in any language.

**Comment by yurishkuro** [comment]
@cijothomas If the API forces you to pass attributes one at a time (whether via varargs or a map) then there will be a proportional growth in cost. My point was that Go SDK already implements so called AttributeSet that, if I understand it correctly, would make the performance the same regardless of the number of attributes used. So if all that we're getting with bound instrument is avoiding 11ns map lookup, yes it's still a 2x improvement, but compared to 200ns for several attributes it's still negligible, and it seems we can do much better without introducing a whole new concept to the spec.

Of course one can argue that the concept of AttributeSet is actually already pretty similar to bound instruments.

Did we have any API proposals of what bound instruments should look like in the API?

**Comment by tsloughter** [comment]
I believe Erlang/Elixir would see a significant boost when it comes to use of integer counters here. Right now measurements have to go through a table lookup and this would remove that need and instead allow direct access to the `counters` reference (a mutable counter builtin to Erlang).

**Comment by utpilla** [comment]
@yurishkuro 

> it seems we can do much better without introducing a whole new concept to the spec.

The motivation for bounded instruments is to avoid a Hashmap lookup when recording a measurement. Hashmap lookup needs both the hash code computation and an equality comparison check for the keys. The more the number of attributes the more would both these operations cost (hash code computation + equality check). I'm not aware of the details of Go's AttributeSet approach but it sounds like they only solved half of the issues with Hashmap lookup. They might have avoided repeated computation of hash code, but they would still incur the cost for equality check.

A bounded instrument would let us avoid these costs entirely by providing us a direct handle to the metric value that needs to be updated.

> So if all that we're getting with bound instrument is avoiding 11ns map lookup, yes it's still a 2x improvement, 

The map lookup cost would not be some constant value for all kinds of measurement recordings. It would be higher for recording higher number of attributes and lengthier key and value strings.

> So what I would really like to understand is whether bound instruments can provide a path to allocation-free (and ideally lock free) counter bumps, and whether the same can be achieved with the existing API.

Allocation-free recording of measurements should be possible even with the existing API. The motivation for bounded instruments is to achieve significantly faster recording of measurements (for highly perf sensitive scenarios) and NOT allocation or lock-free operations.

**Comment by cijothomas** [comment]
> Go SDK already implements so called AttributeSet

`AttributeSet` is not something from the spec, but many languages have it.
Yes sure, such a concept greatly reduces the overall cost, as it can pre-do sort/de-dup/hash-calc etc. on the Attributes ahead of time and re-use. But it won't give the same perf as Bound instruments.

(Note: OTel Rust tried AttributeSet concept only to revert it before shipping : https://github.com/open-telemetry/opentelemetry-rust/pull/1512)

**Comment by jmacd** [comment]
My expectation is that when a bound instrument is used, the map lookup step described above is bypassed. I expect there is an initial step to construct the attribute set, then to perform a map lookup, after which the bound instrument is instantiated, holding a reference to the thing that was looked up. Subsequent operations using the bound instrument should not perform any map lookup-- in the case of a bound counter instrument, I expect a direct reference to an atomic variable, expecting this operation to take O(10ns).

**Comment by jack-berg** [comment]
Some comments from the TC 8/14/24 meeting:

- Useful to accommodate performance sensitive use cases
- Question of whether we would also want to support partially bound instruments
- Question of whether performance improvement would be moot in a world where measurements can be enriched by attribute key-values from context
- There was a previous otep which mentions bound instruments

We would like a reference implementation to show the benefits as a prerequisite to accepting this.

**Comment by cijothomas** [comment]
> Question of whether performance improvement would be moot in a world where measurements can be enriched by attribute key-values from context

True! Bound ones cannot be enriched, it just has the attributes it was bound with. (except the Meter attributes, Resource attributes that are implicitly associated...)

**Comment by cijothomas** [comment]
@jack-berg https://github.com/open-telemetry/opentelemetry-specification/issues/4126#issuecomment-2207011779 has shown the potential gains with bound instruments.. Wondering if that is what TC was looking for?

**Comment by jmacd** [comment]
@cijothomas I apologize for the delay.  I wanted to be sure that we evaluate all the issues connected with this topic.  I definitely support adding bound instruments to the Metric API, and I see it as one of several optimizations that an SDK could offer users to improve the cost of metrics instrumentation.

Instead of a narrow focus on bound instruments, I would like to propose a more aggressive rewrite of the specification that gives each SDK broad permission to extend the metric API with alternatives tailored to specific needs in response to user feedback.  At a high level, what I'm proposing says that the exact mechanics of the metric API do not matter at the specification level, because they are naturally language-specific in the extreme; idioms and practices of each language should govern metric API design.

## Relax the Metrics API specification

The most important aspect of the Metric API/SDK specification, in my opinion, is the terminology and the data model.  As an overview, I would attempt to relax the Metric API specification as follows, roughly:

- The Metric API features instruments that are "compiled" before use, with identifying fields (name, scope name, kind, unit)
- Instruments are either synchronous or asynchronous.
- Every instrument supports exactly one operation, conventially named "Add", "Record", "Set", or "Observe" used to trigger a metric event.
- Each operation (i.e., metric event) is accompanied by exactly one number value and a set of attributes. Note that Metric SDK outcomes are determined by the set of Metric events, not the API mechanics specifically.
- Instruments do not support optionality.  Flags or runtime options that determine semantic impact at the call site are prohibited. Alternative instruments should be crafted instead of overloading existing ones.
- Attribute sets may be compiled into the instrument completely, partiallly, or not, depending on the interface and the circumstances.
- The instruments have standardized naming practice (e.g., a required prefix or suffix or substring in the instrument type name).
- However, instrument names are not unique because multiple forms of the same instrument are encouraged (e.g., optional "bound" instruments).
- SDKs must provide a basic, stable instrument definition corresponding with the first stable Metric API of that SDK.
- SDKs may provide alternative APIs with advanced performance characteristics, provided each alterntive API can be mapped directly to the basic instrument definition (not necessarily efficiently).

I understand why people look to bound instruments -- it's a straightforward, established optimization for Metrics APIs.  However, to use bound instruments requires a tight integration between instrumented code and the Metrics API; users have to inject stateful handles into their code in order to benefit, and the benefits are limited in cases where there is dynamic context.  To me, these detractors make bound instruments a secondary optimization.

I came to the proposal above after considering the thread started here:

https://x.com/juliusvolz/status/1805523949260095672

in which an (IMO) unfair comparision between bound-style and unbound-style interfaces was shared without much context.  One reason the comparison is unfair is that the OTel-Go Metrics API has made concessions in the API with optimization potential similar to bound instruments; the OTel-Go `AttributeSet` (and argument list) can be pre-computed and stored in code to reduce the cost of each event. However, like with bound instruments, this requires code modification to achieve and does not support dynamic instrumentation.

This has been a particular pain point for my internal users of OTel-Go.  We migrated to OTel from a Statsd-flavored Metrics API, where instruments are not compiled and attributes are presented as key-value lists, and there's too much code to update every call site.

In the Statsd-flavored API, attribute sets arrive as a key value list, that is the code path that matters to me.  We should be able to convince ourselves that bound instruments fundamentally can't help with this problem, without my implementing essentially another Metric SDK.  In order for the Statsd-flavored API to benefit from bound instruments, it has to construct a map from a list-of-key-values to bound instrument handles, which is what Metric SDKs are supposed to do.  We should not offer bound instruments so that users can build sophisticated caches of bound instruments, we should provide SDKs with built-in sophisticated caches of bound instruments.

Note that the Prometheus Go client, the library used for the comparison above, implements a sophisticated map from list-of-key-values to bound instruments based on "fingerprinting" the list to avoid memory allocations.  The same could be done in an OTel-Go SDK, [the same is done in the Lighstep Metrics SDK (bypassing the OTel API)](https://github.com/lightstep/otel-launcher-go/blob/main/lightstep/sdk/metric/README.md). Still, for the Stats-flavored API to benefit from fingerprinting, the input has to be list-of-key-value, not `AttibuteSet`; the optimization is literally to avoid constructing the attribute set.  [In other words, the OTel-Go API is fundamentally unable to apply this optimization.](https://github.com/lightstep/otel-launcher-go/issues/446)

To illustrate how I would apply the rules above, some hypothetical examples.

## Hypothetical Example: List-of-key-value alternative API

As described above, there is a lot of existing code that either wasn't written with bound instruments in mind or where there is no way to store a bound instrument, without creating a sophisticated cache keyed by list-of-key-values.  The specification should not prevent a Metrics API from offering multiple input formats through multiple interfaces.

The API could be extended with an alternative list-of-key-value interface, and the SDK would add an optimnized code path based on fingerprinting.  A fallback mechanism would be provided for SDKs that do not recognize the alternative API, which in this case would construct the AttributeSet and call the basic API.

## Hypothetical Example: HistogramWithCount

[There has been a request to add an optional Count argument to the Histogram instrument.](https://github.com/open-telemetry/opentelemetry-specification/issues/3369) While I agree with the motivation for this request, the rules above would require a new instrument to be created instead of overloading the existing histogram.

The API could be extended with an instrument called HistogramWithCount, where the single operation ("Record") takes a second argument ("count").  A fallback mechanism would be provided for SDKs that do not recognize the alternative API, which in this case would repeatedly call a basic Histogram instrument _count_ number of times.

## Hypothetical Example: Struct-tag attributes

The [Segment.io Go Stats library](https://github.com/segmentio/stats) library uses struct tags to convey attribute sets efficiently and idiomatically without memory allocations.  If the OTel-Go community demanded it, and the maintainers agreed, the API could be extended to do this and the SDK could follow an optimized code path for these inputs.  A fallback mechanism would be provided for SDKs that do not recognize the alternative API, which in this case would convert struct-tag metric events into basic metric events.

## Hypothetical Example: Multi-variate metrics API

At times, especially when there is dynamic context, an application may be interested in computing one AttributeSet and then re-using it across a number of metric instruments for a simultaneous observation of a number of values. This was a feature of the OpenCensus Metric API that was de-scoped from OpenTelemetry which has no way to record multi-variate metrics.

While the potential for optimization is good and I would support this proposal, it is probably at odds with the proposed specification changes above.  There is semantic meaning in a multi-variable metric event, and the Metric API has no existing terminology for the pair of Instrument and Value, what OpenCensus and early drafts of OpenTelemetry called "Measurement".  Even though a fallback mechanism could be provided for the multi-variate metric event into many basic metric events, this strays too far from the original specification and would require specification work.

**Comment by ctiller** [comment]
gRPC C++ would love to see a bound instrument API - but we'd need multi-instrument bindings. For a gRPC channel we expect to be exporting 50-100 metrics, and the cost of map lookups per metric is prohibitive, as would be the memory cost of storing pointers to bound metrics on a per metric basis.

**Comment by yurishkuro** [comment]
@ctiller can you elaborate on what kind of solution you have in mind? Your description leaves few options - you don't want to store 50 pointers, but also don't want to do map look ups. I suppose you can have some index-based API where index refers to a vector of metric names stored once.

**Comment by markdroth** [comment]
What we had in mind for gRPC C++ was exactly the "Hypothetical Example: Multi-variate metrics API" case that @jmacd described above.  It's not actually a hypothetical case. :)

**Comment by jsuereth** [comment]
Basically gRPC needs to report many measurements that share the same labelset.  This is something the OpenCensus API used to do better (because you'd report an array of measurements and a set of labels separately in one big batch).

In OpenTelemetry we optimised for the "simple" case of one metric report at a time.  However, multivariate metrics suffer both in the protocol (as @lquerel has been working on) but also in the API.  

I think it's worth us revisiting this issue.

**Comment by ctiller** [comment]
@yurishkuro exactly, some kind of index based design would work. Worth noting that we want to do the map lookup once for a group of metrics at a time, but we may want to modify them at different times. As a concrete use case, for a TCP connection we want to track various latencies and read/write sizes; we want to do the map lookup for a block of "TCP stuff" just once at the establishment of the connection, but over a long period of time update various instruments in that block.

**Comment by lquerel** [comment]
I had started to describe the concept of a metric set (multivariate metrics) as part of the preparatory work for the Schema v2 of the semantic conventions -> https://github.com/open-telemetry/weaver/issues/898 (see description and why metric sets Matter sections).

This is clearly a major need we have at F5 (multivariate metrics for TCP, HTTP (~40 metrics sharing the same 30 attributes), ...) In fact, it was the initial reason why I began contributing to OTEL. In the meantime, I had to adjust my objectives somewhat for various reasons, but we are now closing the loop as @jmacd and @jsuereth mentioned. On one side, a modeling proposal in the semantic conventions, and on the other, an exploration in the OTAP-based Rust engine we are currently developing (https://github.com/open-telemetry/otel-arrow phase 2).

**Comment by dashpole** [comment]
After a bunch of work to improve the Go metrics API and SDK performance, i've found that a bound instrument API would allow me to also fix a [related issue](https://github.com/open-telemetry/opentelemetry-go/issues/7743#issue-3758345145) with how NewSet works in the Go SDK, which was alluded to in @jmacd's post above. Today, creating a NewSet incurs the cost of copying the user-provided attributes for safety. But we really only need a "fingerprint" of the attributes to do a lookup if the attributes have been used with the instrument before. Fixing this would require a minor re-design of our API, and I would like to switch to a bound-instrument pattern for ergonomic reasons as well.  For reference, in a concurrent benchmark, the breakdown of costs for a counter increment are approximately as follows:

1. 45ns: Sort and de-duplicate the provided attributes. 
2. 0-200ns: Compute the hash of attributes (depending on number of attributes).
3. 80-450ns + 1 allocation: Copy the attributes (depending on the number of attributes).
4. 50ns: SDK map lookup on attributes (concurrent).
5. 10 ns: Update the atomic counter.

Today, steps 1, 2, and 3 are part of `attribute.NewSet()`, and 4 and 5 are part of `counter.Add()`.  In the case where attributes are known ahead of time, a counter increment is 60ns.  In the case where attributes are not known ahead of time, a counter increment is ~300 ns (for 3 attributes). The code for performant, dynamic counter increments [is very ugly](https://github.com/open-telemetry/opentelemetry-go/pull/7770/commits/472ef543a52f086ca75d624fc257804894eebf50#diff-d0480e781c9cdba17f26143c510ba9186bb9bc731c2b43acf409e3db52a165c6R83-R96) today.

With bound instruments, in the case where attributes are known ahead of time, a counter increment is 10ns (6x improvement).  In the case where attributes are not known ahead of time, bound instruments would allow us to fix our "fingerprint" issue and skip step 3, which lowers the runtime cost to ~140ns (2x improvement), and **removes the mandatory allocation**.

**Comment by dashpole** [comment]
I also think bound instruments and a multi-variate API could co-exist. I would design it similarly to how the [asynchronous multi-callback API is designed](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#multiple-instrument-callbacks).  For example, in Go:

```golang
// Current API, written in non-performant way.
myCounter.Add(ctx, 1, metric.WithAttributes(attribute.String("foo", "bar")))
myHistogram.Record(ctx, 1, metric.WithAttributes(attribute.String("foo", "bar")))

// Example with bound instruments, which is much more performant.
myCounter.WithAttributes(attribute.String("foo", "bar")).Add(ctx, 1)
myHistogram.WithAttributes(attribute.String("foo", "bar")).Record(ctx, 1)

// Example with multi-variate instruments, and "bound meters"
meter.WithAttributes(attribute.String("foo", "bar")).MultiMeasure(ctx,
    func(m metric.Measurer) {
        m.MeasureInt64(myCounter, 1)
        m.MeasureInt64(myHistogram, 1)
    }, myCounter, myHistogram)
```

**Comment by jmacd** [comment]
This looks great! I agree with

>  Fixing this would require a minor re-design of our API, and I would like to switch to a bound-instrument pattern for ergonomic reasons as well. 

We are at the point where bound instruments are good for multiple reasons.

> meter.WithAttributes(attribute.String("foo", "bar")).MultiMeasure(...)

Nice!

**Comment by cijothomas** [comment]
> 3\. 80-450ns + 1 allocation: Copy the attributes (depending on the number of attributes).

Thanks @dashpole ! The above is something .NET addresses by using thread-local. There is still copy cost, but that can be fixed with https://github.com/open-telemetry/opentelemetry-dotnet/issues/5777. Rust already avoids this copy as it can do a look-up with user-provided attributes (even if they live of stack).


The 10ns for bound-instrument update matches our [observation](https://github.com/open-telemetry/opentelemetry-specification/issues/4126#issuecomment-2207011779) as well!

**Comment by cijothomas** [comment]
> 30ns: Sort and de-duplicate the provided attributes.

This is surprisingly low! Sorting/de-dup is much more expensive in .NET/Rust! (~200ns for 5 attributes). I need to see how go can do it for 30ns! 

Was that for 1 or 2 attributes?

**Comment by dashpole** [comment]
I re-ran my custom benchmark, and it should have been 45 ns for 3 attributes.

For 1, 2, 3, 4, 5 attributes, it takes 8 ns, 27 ns, 45 ns, 64 ns, 82 ns.

**Comment by cijothomas** [comment]
https://github.com/open-telemetry/opentelemetry-rust/pull/3392 OTel Rust PR exploring bound instruments. Benchmarks confirm the expected numbers - Bound counters are 1-5 ns only, Bound histograms 5-10 ns. Still in review stage, and some open questions exist. If anyone wants to take a peek and share feedback, please do.

**Comment by jack-berg** [comment]
Are we at the point where we can label this as ready with sponsor? I believe we are and I am willing to sponsor. 

As for the multivariate metrics API piece: I see a good amount of support here for that, and I support as well. However, I propose we tackle the two problems independently, and solve bound instruments before multivariate metrics API, since multivariate metrics API seems useful when built on top of bound instruments.

**Comment by dashpole** [comment]
We discussed this at the Go SIG in depth, and there **was not** agreement that we would want to implement this if it is added to the specification. The main concerns from my recollection is that it duplicated existing API surface to pass attributes to measurements, and would introduce a new pattern. I'm still happy to participate in the design.

**Comment by cijothomas** [comment]
> Are we at the point where we can label this as ready with sponsor? I believe we are and I am willing to sponsor.
> 
> As for the multivariate metrics API piece: I see a good amount of support here for that, and I support as well. However, I propose we tackle the two problems independently, and solve bound instruments before multivariate metrics API, since multivariate metrics API seems useful when built on top of bound instruments.

Before we proceed, we'd need to see if we have enough interest from languages to spend the effort in Spec. (Usually, we want 3 implementations for stable spec, but if we don't have that, then this will remain forever stuck in experimental..)

(I can confirm OTel Rust 100% wants this. We prefer to have it come via Spec, but if there isn't enough appetite, we can offer it as a additional Rust specific feature)

**Comment by jack-berg** [comment]
Additionally, java wants this.

Based on comment in this thread, it sounds like go is interested.

@cijothomas do you know if .NET is interested?

@tsloughter do you know if erlang is interested?

**Comment by cijothomas** [comment]
@martincostello @rajkumar-rangaraj @alanwest Can you comment if OTel .NET has active interest in bound instruments? it was discussed a while ago, but no recent activity/discussion that I can find.

**Comment by martincostello** [comment]
I'm not aware of any specific interest in this area for .NET.

**Comment by verdie-g** [comment]
Some interest here https://github.com/dotnet/extensions/issues/6513 :)

**Comment by jmacd** [comment]
I think we are all ready for this. I want to be sure the languages each feel they have a choice, but I think you have at least 3 languages with an interest. If it requires a major redesign of the SDK, I can understand why this would be an obstacle. 

I have one historical example to share, which is the [Lightstep Metrics SDK](https://github.com/lightstep/otel-launcher-go/blob/main/lightstep/sdk/metric/README.md) which I authored at a previous company. This SDK was designed with bound instruments in mind, and it really does add substantial complexity. The synchronous measurement APIs were implemented as a sequence of effectively `inst := provider.bind(); defer inst.unbind(); inst.operate(...)` . Still, I can speak for the Collector to say that there is an interest in making metrics cost less and there are structures in place to bind instruments.

As a final concern, I think we should say something about the crossover between this and #4931. There has been a discussion, once a PR https://github.com/open-telemetry/opentelemetry-specification/pull/4318, on the `MeasurementProcessor` concept. The concept is more broad than this, for example a MeasurementProcessor could change the measured value and the units, maybe, but the only reason I've had for MeasurementProcessor personally would be accomplished with Context-scoped attributes. IOW this gives maybe 90% of the value of MeasurementProcessor because it means count everything by configurable dimensions.

**Comment by jmacd** [comment]
^^^ @jack-berg @carlosalberto 

This is an important point we should cover when designing bound instruments: the point of bound instruments is to make measurements really fast, but do we gain from binding instruments when there are context-scoped attributes? Do we support "partial binding"? I'm not sure anyone has implemented this. @bogdandrutu has made statements on this topic!

**Comment by jack-berg** [comment]
> but I think you have at least 3 languages with an interest.

"accepting" in this context is just an indication that we're going to work on it. It has a long way to go before its a merged stable part of the spec, with plenty of opportunities to discover issues and derail along the way 😉. Let's get that process started though. I'm reasonably confident we can work through the issues and land something useful.

> If it requires a major redesign of the SDK, I can understand why this would be an obstacle.

I don't anticipate a major redesign of the java SDK. Of course I won't know for sure until I build it, but I have a pretty good idea of how it would fit in, and think it would mostly be additive work rather than major refactoring.

> The synchronous measurement APIs were implemented as a sequence of effectively

The java SDK already does something like this today even without bound instruments, which may be why it seems like a natural fit 😁 

> This is an important point we should cover when designing bound instruments: the point of bound instruments is to make measurements really fast, but do we gain from binding instruments when there are context-scoped attributes? 

I think no. And this seems fundamental. Bound instrument use is applicable in a subset of use cases where the set of distinct attribute sets is known up front (or at compile time if you like). Context scope attributes break this and revert the resolved attributes to be dynamic at runtime. Sure we can make the API work, but the performance benefit will go away.

To me this is fine. Bound instruments have always been a bit niche - most of the popular metrics in semconv require computing attributes at runtime - but they really matter to those niche use cases which are very performance sensitive. If you need the performance of bound instruments, don't use context scoped attributes. 

> Do we support "partial binding"?

I would be surprised if there was enough perf gain to warrant the complexity of partial binding. Interested in other POVs though - maybe I'm missing something.

**Comment by jack-berg** [comment]
Java prototype: https://github.com/open-telemetry/opentelemetry-java/pull/8314

Modest to large gains across the board, with larger gains for cases with reduced contention and cumulative temporality, where the map lookup represents a larger share of the time to record. (In java, we have a variety of optimizations in place, including separating the cumulative and delta metric storage codepaths. Since delta requires additional coordination between record and collect threads, cumulative is already very fast, and so removing something like a hashmap lookup makes up a bigger portion of the record time)


| Threads | Temporality | Cardinality | Instrument | Bound=false (ops/s) | Bound=true (ops/s) | Δ ops/s | Δ % |
|---|---|---|---|--:|--:|--:|--:|
| 1 | DELTA | 1 | COUNTER_SUM | 118,208,794 | 130,599,635 † | +12,390,841 | +10.5% |
| 1 | DELTA | 1 | UP_DOWN_COUNTER_SUM | 102,544,568 | 129,911,312 † | +27,366,744 | +26.7% |
| 1 | DELTA | 1 | GAUGE_LAST_VALUE | 36,996,170 | 48,977,805 | +11,981,635 | +32.4% |
| 1 | DELTA | 1 | HISTOGRAM_EXPLICIT | 58,932,937 | 122,397,793 | +63,464,856 | **+107.7%** |
| 1 | DELTA | 1 | HISTOGRAM_BASE2_EXPONENTIAL | 43,456,120 | 44,312,699 | +856,579 | +2.0% |
| 1 | DELTA | 128 | COUNTER_SUM | 94,602,944 | 114,178,751 † | +19,575,807 | +20.7% |
| 1 | DELTA | 128 | UP_DOWN_COUNTER_SUM | 99,861,595 | 114,805,474 † | +14,943,879 | +15.0% |
| 1 | DELTA | 128 | GAUGE_LAST_VALUE | 30,515,600 | 36,287,352 | +5,771,752 | +18.9% |
| 1 | DELTA | 128 | HISTOGRAM_EXPLICIT | 68,355,768 | 89,869,366 | +21,513,598 | +31.5% |
| 1 | DELTA | 128 | HISTOGRAM_BASE2_EXPONENTIAL | 40,758,132 | 46,997,748 | +6,239,616 | +15.3% |
| 1 | CUMULATIVE | 1 | COUNTER_SUM | 165,529,273 | 216,674,644 | +51,145,371 | +30.9% |
| 1 | CUMULATIVE | 1 | UP_DOWN_COUNTER_SUM | 167,603,291 | 216,483,041 | +48,879,750 | +29.2% |
| 1 | CUMULATIVE | 1 | GAUGE_LAST_VALUE | 48,499,078 | 75,562,700 | +27,063,622 | +55.8% |
| 1 | CUMULATIVE | 1 | HISTOGRAM_EXPLICIT | 98,273,713 | 134,302,636 | +36,028,923 | +36.7% |
| 1 | CUMULATIVE | 1 | HISTOGRAM_BASE2_EXPONENTIAL | 46,687,571 | 51,549,345 | +4,861,774 | +10.4% |
| 1 | CUMULATIVE | 128 | COUNTER_SUM | 88,309,144 | 214,865,227 | +126,556,083 | **+143.3%** |
| 1 | CUMULATIVE | 128 | UP_DOWN_COUNTER_SUM | 97,199,593 | 205,305,518 | +108,105,925 | **+111.2%** |
| 1 | CUMULATIVE | 128 | GAUGE_LAST_VALUE | 102,375,233 | 204,971,971 | +102,596,738 | **+100.2%** |
| 1 | CUMULATIVE | 128 | HISTOGRAM_EXPLICIT | 75,716,082 | 113,538,456 | +37,822,374 | +49.9% |
| 1 | CUMULATIVE | 128 | HISTOGRAM_BASE2_EXPONENTIAL | 44,206,432 | 48,689,276 | +4,482,844 | +10.1% |
| 4 | DELTA | 1 | COUNTER_SUM | 15,865,839 | 18,124,749 | +2,258,910 | +14.2% |
| 4 | DELTA | 1 | UP_DOWN_COUNTER_SUM | 18,307,609 | 17,372,346 | -935,263 | -5.1% |
| 4 | DELTA | 1 | GAUGE_LAST_VALUE | 12,223,215 | 17,542,726 | +5,319,511 | +43.5% |
| 4 | DELTA | 1 | HISTOGRAM_EXPLICIT | 12,133,325 | 12,767,563 | +634,238 | +5.2% |
| 4 | DELTA | 1 | HISTOGRAM_BASE2_EXPONENTIAL | 10,102,515 | 10,626,084 | +523,569 | +5.2% |
| 4 | DELTA | 128 | COUNTER_SUM | 76,699,353 | 68,831,065 | -7,868,288 | -10.3% |
| 4 | DELTA | 128 | UP_DOWN_COUNTER_SUM | 74,165,415 | 65,990,906 | -8,174,509 | -11.0% |
| 4 | DELTA | 128 | GAUGE_LAST_VALUE | 50,121,436 | 50,574,389 | +452,953 | +0.9% |
| 4 | DELTA | 128 | HISTOGRAM_EXPLICIT | 60,837,029 | 59,375,391 | -1,461,638 | -2.4% |
| 4 | DELTA | 128 | HISTOGRAM_BASE2_EXPONENTIAL | 61,647,646 | 57,771,684 | -3,875,962 | -6.3% |
| 4 | CUMULATIVE | 1 | COUNTER_SUM | 72,849,913 | 74,565,813 | +1,715,900 | +2.4% |
| 4 | CUMULATIVE | 1 | UP_DOWN_COUNTER_SUM | 75,558,659 | 55,029,449 | -20,529,210 | -27.2% |
| 4 | CUMULATIVE | 1 | GAUGE_LAST_VALUE | 28,552,341 | 28,218,133 | -334,208 | -1.2% |
| 4 | CUMULATIVE | 1 | HISTOGRAM_EXPLICIT | 15,952,626 | 21,438,633 | +5,486,007 | +34.4% |
| 4 | CUMULATIVE | 1 | HISTOGRAM_BASE2_EXPONENTIAL | 16,650,679 | 17,252,153 | +601,474 | +3.6% |
| 4 | CUMULATIVE | 128 | COUNTER_SUM | 114,706,054 | 123,732,233 | +9,026,179 | +7.9% |
| 4 | CUMULATIVE | 128 | UP_DOWN_COUNTER_SUM | 110,953,644 | 122,402,433 | +11,448,789 | +10.3% |
| 4 | CUMULATIVE | 128 | GAUGE_LAST_VALUE | 101,753,088 | 108,879,557 | +7,126,469 | +7.0% |
| 4 | CUMULATIVE | 128 | HISTOGRAM_EXPLICIT | 77,679,812 | 83,538,482 | +5,858,670 | +7.5% |
| 4 | CUMULATIVE | 128 | HISTOGRAM_BASE2_EXPONENTIAL | 75,555,258 | 79,784,244 | +4,228,986 | +5.6% |

⚠️ Several rows with `isBound=true` have very high variance (>20%) — marked with † — treat those deltas with caution.

In my opinion, while the degree of the improvement is situational, given the large improvement on the top end, the modest improvement across the board, and the minorness of the UX impact, it would be easy for me to recommend using this API to users if their use case allows it.
